[c10d][Doc] Add a flight recorder tutorial (#3040)
pytorch/tutorials: prototype_source/flight_recorder_tutorial.rst

- **Resource Constraints:** If the system running the job does not have enough computational resources (such as CPU, GPU, or memory), the job might not be able to proceed.

- **Network Issues:** In a distributed training setup, different parts of the model or data may be processed on different devices. If there are network issues, communication between these devices may be disrupted, causing the job to get stuck.

- **Software Bugs or Errors:** Errors in the training code or the underlying libraries and frameworks can also cause a job to get stuck.

- **Synchronization Issues:** In distributed training, different parts of the computation are often run in parallel and need to be synchronized at certain points. If this synchronization fails, the job can get stuck. For example, a deadlock can occur if one or more ranks fail to join a collective while the remaining ranks have joined. This results in an indefinite wait for the job to progress.

Flight Recorder, as the name suggests, captures diagnostics information as collectives run. The captured diagnostic

> information is used to help root cause issues when jobs get stuck.

Reviewer: ```suggestion
information is used to help identify the root causes of issues when jobs become stuck.
```

> There are two core parts to Flight Recorder.

Reviewer: ```suggestion
Flight Recorder consists of two core parts: 
```

Enabling Flight Recorder
------------------------
There are two required environment variables to get the initial version of Flight Recorder working.

- ``TORCH_NCCL_TRACE_BUFFER_SIZE = (0, N)``: Setting ``N`` to a positive number enables collection.
     ``N`` represents the number of entries that will be kept internally in a circular buffer.
     We recommended to set this value at *2000*.
- ``TORCH_NCCL_DUMP_ON_TIMEOUT = (true, false)``: Setting this to ``true`` will write out diagnostic files to disk on job timeout.
     If enabled, there will be one file per rank output in the job's running directory.

**Optional settings:**

- ``TORCH_NCCL_TRACE_CPP_STACK = (true, false)``: Setting this to true enables C++ stack traces to be captured in Flight Recorder.
     C++ stack traces can be useful in providing the exact code path from a PyTorch Python call down to the primitive
     C++ implementation. Also see ``TORCH_SYMBOLIZE_MODE`` in additional settings.
- ``TORCH_NCCL_ENABLE_TIMING = (true, false)``: Setting this to ``true`` will enable additional cuda events at the start of each collective and

>      records the *duration* of each collective. This may incur some CPU overhead. In the collected data, the

Reviewer: ```suggestion
  records the *duration* of each collective. This may incur some CPU overhead. In the collected data, the
```

>      *duration* field indicates how long each collective took to execute.

Reviewer: ```suggestion
  *duration* field indicates how long each collective took to execute.
```
